activerecord/lib/active_record/associations/builder/has_one.rb

```diff
@@ -32,26 +32,20 @@ def self.define_validations(model, reflection)
       end
     end
 
-    def self.touch_record(o, name, touch)
-      record = o.send name
+    def self.touch_record(record, name, touch)
+      instance = record.association(name).target
```

Comment:
I've identified actual change in this PR is only this line.
At a glance this seems similar with #31010, but I'll take a look deeply tomorrow since I have no time today.

Reply:
@kamipo sorry, I should've mentioned that this was the main change. And you're right it's similar to that. It's generally a problem when accessing records within a create transaction when the reference isn't set (e.g. it would have to go to the db where nothing has been created yet).

Do you see any major problems with this approach? Otherwise I'll move forward with this.

Reply:
I've indentified what is happened. Looks like that `stale_target?` for has_one doesn't properly work after association target is loaded. This change would break has_one touching if association target is not loaded yet since `target` doesn't cause loading association target:

```diff
diff --git a/activerecord/test/cases/associations/has_one_associations_test.rb b/activerecord/test/cases/associations/has_one_associations_test.rb
index 2ee260ef2b..f17c2580e7 100644
--- a/activerecord/test/cases/associations/has_one_associations_test.rb
+++ b/activerecord/test/cases/associations/has_one_associations_test.rb
@@ -712,6 +712,18 @@ def test_has_one_with_touch_option_on_create
     }
   end
 
+  def test_foo
+    DrinkDesignerWithPolymorphicTouchChef.create(chef: Chef.new)
+    designer = DrinkDesignerWithPolymorphicTouchChef.last
+
+    # UPDATE "drink_designers" SET "name" = ? WHERE "drink_designers"."id" = ?
+    # SELECT "chefs".* FROM "chefs" WHERE "chefs"."employable_id" = ? AND "chefs"."employable_type" = ? LIMIT ?
+    # UPDATE "chefs" SET "updated_at" = ? WHERE "chefs"."id" = ?.
+    assert_queries(3) {
+      designer.update(name: "foo")
+    }
+  end
+
   def test_polymorphic_has_one_with_touch_option_on_create_wont_cache_assocation_so_fetching_after_transaction_commit_works
     assert_queries(3) {
       chef = Chef.create(employable: DrinkDesignerWithPolymorphicTouchChef.new)
diff --git a/activerecord/test/schema/schema.rb b/activerecord/test/schema/schema.rb
index b6c0ae0de2..5e4893946a 100644
--- a/activerecord/test/schema/schema.rb
+++ b/activerecord/test/schema/schema.rb
@@ -1062,6 +1062,7 @@
   create_table :cake_designers, force: true do |t|
   end
   create_table :drink_designers, force: true do |t|
+    t.string :name
   end
   create_table :chefs, force: true do |t|
     t.integer :employable_id
@@ -1069,6 +1070,7 @@
     t.integer :department_id
     t.string :employable_list_type
     t.integer :employable_list_id
+    t.timestamps
   end
   create_table :recipes, force: true do |t|
     t.integer :chef_id
```